### ISSUE TITLE
Type of subexpressions in file

### ISSUE THREAD

**Issue by NightRa** [idris-lang/Idris-dev#2608]
It would be really nice if we could inspect types of non top-level definitions, as in ghc-mod type:
Given a file and a cursor location, give the type of the expression.
This would be useful for editors - to be able to see types of whatever we mark.

**Comment by david-christiansen** [comment]
I agree! It would also be nice to export parse trees to editors, to implement things like structured-haskell-mode. Implementing this right now would be a little bit tedious, largely because we don't track source spans exactly enough in the parser. After that, it could use the same highlighting infrastructure that's used to put semantic colors on source code to also send editors type information. If you want to have a go, I'll walk you through what needs doing!

**Comment by ben-schulz** [comment]
Unless @NightRa or someone else has already claimed this task, I might be interested in picking it up.  What would it entail?

**Comment by NightRa** [comment]
I'm very interested in doing this one. I consider maybe taking it as a mini
project for my undergrad project.

**Comment by ben-schulz** [comment]
I'll leave it to you then -- and I look forward to seeing how it turns out!

**Comment by david-christiansen** [comment]
Whoever ends up taking it on, the code to look at is in `src/Idris/Elab/Term.hs` and `src/Idris/ParseExpr.hs` and `src/Idris/AbsSyntaxTree.hs`.

`Term.hs` contains the term elaborator (which is the cursive-E and cursive-P operators in Edwin's paper on the elaborator from 2013). The term elaborator is what converts desugared high-level terms (datatype `PTerm`) into the core language expressions (datatype `TT Name`, aliases `Type` and `Term`). Some of these `PTerm` constructors contain information about precise source locations of names and other constructs. When the term elaborator encounters these precisely-annotated AST nodes, then it emits information about that precise source span with the `highlightSource` function. This function takes a source span (datatype `FC`) and a semantic annotation for that span (datatype `OutputAnnotation`). Probably the right thing to do here is to add another constructor to `OutputAnnotation` along the lines of `SourceCodeType :: Type -> OutputAnnotation`.

`PTerm` is defined in `AbsSyntaxTree.hs`. It has two kinds of source span, and I should probably wrap one of them in a newtype because it's a bit confusing that both of them are `FC` right now - doing this might be a good first step in this project! Some of them represent precise character-accurate source spans (such as on `PRef`), others are general locations that put an error in basically the right spot. To make this work, you'll have to make more of them precise - at least the one for `PApp` (function application). You may also need a new constructor `DesugaredFrom :: FC -> PTerm -> PTerm` that represents that its second argument was desugared from a term whose precise location is its first argument. This is because the elaborator in `Term.hs` never sees things like `if-then-else-`, but it might be nice to be able to get its type.

Finally, the tedious part of this job would be in `ParseExpr.hs`. There, you'll need to figure out how to record the end of each read token, and make sure to save that span rather than the span including trailing whitespace when you save the new precise locations for things like `PApp` and `PIfThenElse`.

Please let me know if any of this was unclear or if you need help with some of the code!

I can also help out with the Elisp once the Haskell side is ready - it should be something like this:
```
(defvar-local idris-highlighted-source-expression-overlay nil)
(defun idris-clear-expression-type ()
  (when idris-highlighted-source-expression
    (overlay-put idris-highlighted-source-expression-overlay 'face nil)))
(defun idris-show-expression-type (point)
  (let* ((overlays (overlays-at point))
         (in-order (SORT THE OVERLAYS BY SIZE, SMALLEST FIRST))
         (remaining (DROP THE OVERLAYS UNTIL idris-highlighted-source-expression IS THE HEAD))
         (next-overlay (if overlays (if remaining (car remaining) (car overlays)) nil)))
    (if (null next-overlay)
        (error "No typed expression at here")
      (setq idris-highlighted-source-expression next-overlay)
      (overlay-put next-overlay 'face 'idris-typed-source-expression-face)
      (message (overlay-get next-overlay 'idris-expression-type)))))
```
and then perhaps update Eldoc to also show the expression type instead of the identifier type when there's an active highlighted source expression.
  
```

**Comment by NightRa** [comment]
@ben-schulz You may take it for yourself :)
I'm doing something else in the end.
